OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,266 @@
+			widget.GetText = () => item;
+			return widget;
+		}
+
+		void AddChatLine(string text)
+		{
+			Log.Write("irc", text);
+
+			var scrolledToBottom = historyPanel.ScrolledToBottom;
+
+			var newChild = MakeLabelWidget(historyTemplate, text);
+			historyPanel.AddChild(newChild);
+
+			if (scrolledToBottom)
+				historyPanel.ScrollToBottom(smooth: true);
+		}
+
+		bool EnterPressed()
+		{
+			if (!inputBox.Text.Any())
+				return true;
+
+			var text = inputBox.Text;
+			inputBox.Text = "";
+
+			if (text.StartsWith("/"))
+				AddChatLine("Unknown command");
+			else
+			{
+				AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), client.Nickname, text));
+				client.SendMessage(SendType.Message, "#" + Game.Settings.Irc.Channel, text);
+			}
+
+			return true;
+		}
+
+		public void OnError(object sender, ErrorEventArgs e)
+		{
+			AddChatLine("Errror: ".F(e.ErrorMessage));
+			Log.Write("irc", "Error: {0}".F(e.ErrorMessage));
+		}
+
+		public void OnRawMessage(object sender, IrcEventArgs e)
+		{
+			Log.Write("irc", e.Data.RawMessage);
+
+			if (e.Data.Type == ReceiveType.ChannelMessage)
+				Game.RunAfterDelay(client.Lag.Milliseconds * 5, () =>
+					AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), e.Data.Nick, e.Data.Message)));
+
+			if (e.Data.Channel == null)
+				return;
+
+			var channel = client.GetChannel(e.Data.Channel);
+			if (channel == null || !channel.IsSycned)
+				return;
+
+			if (e.Data.Type == ReceiveType.Part)
+			{
+				AddChatLine("{0} left {1}.".F(e.Data.Nick, e.Data.Channel));
+				Game.RunAfterDelay(client.Lag.Milliseconds * 5, () => SyncChannelUsers(channel));
+			}
+
+			if (e.Data.Type == ReceiveType.Join)
+			{
+				AddChatLine("{0} joined {1}.".F(e.Data.Nick, e.Data.Channel));
+				Game.RunAfterDelay(client.Lag.Milliseconds * 5, () => SyncChannelUsers(channel));
+			}
+		}
+
+		public void OnJoin(object sender, JoinEventArgs e)
+		{
+			var channel = client.GetChannel(e.Data.Channel);
+			if (channel == null || !channel.IsSycned)
+				AddChatLine("Joined channel {0}.".F(e.Channel));
+		}
+
+		public void OnChannelActiveSynced(object sender, IrcEventArgs e)
+		{
+			var channel = client.GetChannel(e.Data.Channel);
+
+			AddChatLine("{0} users online".F(channel.Users.Count));
+
+			if (!string.IsNullOrEmpty(channel.Topic))
+				AddChatLine("*** Topic: {0}".F(channel.Topic));
+
+			SyncChannelUsers(channel);
+		}
+
+		void SyncChannelUsers(Channel channel)
+		{
+			nicknamePanel.RemoveChildren();
+
+			foreach (DictionaryEntry user in channel.Users)
+			{
+				var channeluser = (ChannelUser)user.Value;
+				var prefix = channeluser.IsOp ? "@" : channeluser.IsVoice ? "+" : "";
+				var newChild = MakeLabelWidget(nicknameTemplate, prefix + channeluser.Nick);
+				nicknamePanel.AddChild(newChild);
+			}
+		}
+
+		public void OnConnected(object sender, EventArgs e)
+		{
+			AddChatLine("Connected.");
+
+			AddChatLine("Loggin in...");
```

Comment:
As I mentioned previously, this message is unnecessary (and also spelled incorrectly fwiw)